### DISCUSSION TITLE
Cursorless for 3D VR visual programming language

### DISCUSSION THREAD

**Discussion by VestigeJ** [cursorless-dev/cursorless#2059] [Ideas]
Title says a lot but i've been contemplating the talk given by Bret Victor titled "The Future of Programming" 

There are a growing number of opportunities to re-think methods for programming. In my case i'm imagining a rendered 3D object representation of code where the primitives are functions all in pure WASM under the hood. Private functions where typical structural language like "loop", "case", etc are represented in the 3D rendered object. Features like shape, facets, length, width, height, color, etc can be tied to various or specific attributes of the function. Number of public vs private functions, number of variables within, maybe time consumed by the cpu or color for the size of the memory footprint. The overarching goal would be the removal of text exception for numbers at least on the programming side. Has anyone else considered something like this recently?

**Comment by pokey** [comment]
Cool idea! Shall we chat about it in one of our Discord meet-ups? If you drop into the Slack from the talon voice homepage and look for the cursorless channel we can go from there

One thing I'll say is that structured visualisations are awesome, but I'm hesitant to move fully away from the malleability / flexibility of a textual representation for code. I talk about some of the trade-offs there towards the end of my strange loop talk (in the "smart vs dumb" section)

**Reply by conartist6** [reply]
I'd love to have a more complete conversation with you about "smart vs dumb" because when you have appropriate tools I still don't think it's very hard to stay smart

**Comment by VestigeJ** [comment]
There are definitely some fore-runners in this arena for inspiration - https://www.inkandswitch.com/inkbase/

https://www.roadtovr.com/watch-logix-impressive-multi-user-visual-programming-interface-vr/

**Reply by pokey** [reply]
Thanks for the links! Ink and switch crew def doing lots of cool stuff. Let's chat sometime in the discord!

**Comment by VestigeJ** [comment]
So I have done a bit more research but I did end up finding a talon script setup for Uiua which is an Array Programming Language written in Rust that I think could be a great starting point. The more open ended part is still figuring out how to represent code in a three dimensional space with the intent of easier intuition. I honestly think this is where array programming languages could shine as they are more inherently dimensional and succinct.  https://youtu.be/-VOxSNyaLTg?si=liP5HBVzDfFNq3LH

**Comment by conartist6** [comment]
I think about stuff like this a lot, but the difference for me is that I don't think we're talking about one language, but a way to use tactile, semantic VR-able interfaces to edit code written in any language. After all, a solution that only solves the problem for one language will hardly change the future of programming.

**Reply by conartist6** [reply]
To expand on that: moving all code editing to the semantic layer would be a pretty major win for editing languages that are only defined at the semantic layer, which is to say they lack traditional syntax.

**Comment by VestigeJ** [comment]
Pokey introduced me to this community which is most likely a better place for future conversations of this nature - going to close this with this link back there. 

https://futureofcoding.org/community
https://join.slack.com/t/futureofcoding/shared_invite/zt-2bxrongqa-21C1dpgatX8srAGJKWtjYA

**Reply by pokey** [reply]
Glad that community is seeming like a good fit! I'm going to keep this discussion open though because I think there are aspects of using Cursorless in VR that may be interesting to discuss
